userName/src/main/java/com/example/userName/NameController.java

```diff
@@ -0,0 +1,56 @@
+package com.example.userName;
+
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+import org.springframework.web.util.UriComponentsBuilder;
+
+import java.net.URI;
+import java.util.Map;
+
+
+@RestController
+public class NameController {
+    private final UserInfo userInfo;
+
+    private NameController(UserInfo userInfo) {
+        this.userInfo = userInfo;
+    }
+
+
+    @GetMapping("/userinfo")
+    public String getUserinfo(@RequestParam(value = "name", defaultValue = "name") String UserName, @RequestParam(value = "name", defaultValue = "birthdays") String UserBirthDay) {
+        return "私の名前は" + userInfo.service(UserName) + "です。生年月日は" + userInfo.userBirthDays(UserBirthDay) + "です。";
+
+    }
+
+    @GetMapping("/birthday")
+    public String getBirthDay(@RequestParam(value = "name", defaultValue = "birthdays") String UserBirthDay) {
+        return "生年月日は" + userInfo.userBirthDays(UserBirthDay) + "です。";
+    }
+
+    @GetMapping("/names")
+    public String getNames(@RequestParam(value = "name", defaultValue = "name") String UserName) {
+        return "私の名前は" + userInfo.service(UserName) + "です。";
+    }
+
+    @PostMapping("/names")
+    public ResponseEntity<Map<String, String>> create(@RequestBody CreateForm form, UriComponentsBuilder uriBuilder) {
+        // 登録処理は省略 <>はジェネリクス→型に依存しない。
+        String name = form.getName();
+        URI url = UriComponentsBuilder.fromUriString("http://localhost:8080")
```

Comment:
引数にUriComponentsBuilderをとるならば、`"http://localhost:8080"`をハードコードしなくてもレスポンスヘッダーのLocationに正しくホストが設定されるはずです。
```suggestion
    public ResponseEntity<Map<String, String>> create(@RequestBody CreateForm form, UriComponentsBuilder uriBuilder) {
        // 登録処理は省略 <>はジェネリクス→型に依存しない。
        String name = form.getName();
        URI url = uriBuilder
```

Reply:
わかりました。